Fix original object mutation on patch retry

liggitt opened the pull request:
When applying a patch, the original state of the object and patch are captured so that retries can determine if an overlapping update was made to the object, and the patch API call should abort with a conflict.

However, the `strategicPatchObject -> applyPatchToObject -> StrategicMergeMapPatch -> mergeMap` call mutates both the original object map and the patch map, making them unsuitable for reusing for subsequent calls.

We saw this in a [downstream test](https://github.com/openshift/origin/blob/master/test/integration/patch_test.go) that exercises patch conflict retries, where the data being submitted in a patch was showing up in the original object data map.

Since mergeMap *also* mutates the patch map passed in (deletes patch directives as it encounters them), we *also* cannot reuse the patch map in `applyPatchToObject` once it has been used.

This PR:
* Builds `originalObjMap` separate from the initial patch application for later use in conflict detection
* Changes the `originalPatchMap` to a helper that returns a fresh map from original sources
* Adds an integration test that exercises retry of non-overlapping patches with patches containing `$patch` directives

```release-note
Fix incorrect conflict errors applying strategic merge patches to resources.
```

liggitt committed:
Fix original object mutation on patch retry

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43871)
<!-- Reviewable:end -->

liggitt commented:
cc @kubernetes/sig-api-machinery-pr-reviews @sttts @ncdc

liggitt commented:
@wojtek-t @deads2k PTAL

liggitt commented:
Tagging for cherry-pick since this was a regression in 1.6

sttts commented:
Looks like this was failed communication: here is another one: #43902, basically the same thing. We can choose one.

liggitt commented:
The patch map also gets mutated. This fixed and tests that issue as well

deads2k commented:
I think this does what it sets out to do, but the code in this area is really sticky to read.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -86,21 +86,21 @@ func strategicPatchObject(
 	versionedObj runtime.Object,
-) (originalObjMap map[string]interface{}, patchMap map[string]interface{}, retErr error) {
-	originalObjMap = make(map[string]interface{})
+) error {
```
I just removed the return values here since they are unsuitable for use after having used them to apply the patch

wojtek-t commented on test/integration/apiserver/patch_test.go:
```diff
@@ -0,0 +1,114 @@
+	}
+	wg.Wait()
+
+	if successes < handlers.MaxRetryWhenPatchConflicts {
```
I think I don't understand this. We have 2N concurrent goroutines trying to do a patch. Why at least N of them has to succeed? Can you explain it (and add comment about it)?

sttts replied:
https://github.com/kubernetes/kubernetes/pull/43902/files#diff-a98175d31a994ac0386ef984acfe13f5R111

@liggitt wants to copy that into his patch. Took some time to get the reasoning there.

wojtek-t replied:
BTW - I just looked into the other PR, and I think I like the approach from #43902 better. So maybe we should proceed with that one and close this one?
@liggitt @sttts  - thoughts?

liggitt replied:
I can rename the function here as well. That one doesn't address the patch map mutation issue

wojtek-t replied:
I see. But the test is still passing with the other PR, which means it's not really testing everything, right?

Anyway - I'm fine with proceeding with this one too - just want to understand it.

liggitt replied:
> the test is still passing with the other PR

the test in the other PR doesn't include the patch directives... the test in this PR would fail against the fix in the other PR

wojtek-t replied:
OK - sorry for confusion then.

wojtek-t replied:
Can you then please add a comment (similar than the one from the other PR) and I will lgtm this PR.

wojtek-t reviewed (commented):
Sorry for long delay - I added some comment to a test.

liggitt commented:
added comment at https://github.com/kubernetes/kubernetes/pull/43871/files#diff-a98175d31a994ac0386ef984acfe13f5R60

I added a mutation warning to the StrategicMergePatch godoc for now, rather than rename

wojtek-t commented:
> I added a mutation warning to the StrategicMergePatch godoc for now, rather than rename

That's fine. Thanks.

/lgtm

liggitt commented:
@k8s-bot kops aws e2e test this
